amxmodx/tempent.cpp

```diff
@@ -0,0 +1,1504 @@
+		return 0;
+
+	cell *position = get_amxaddr(amx, params[arg_position]);
+
+	MESSAGE_BEGIN(get_msg_destination(index, params[arg_reliable] != 0), SVC_TEMPENTITY, NULL, index ? TypeConversion.id_to_edict(index) : NULL);
+		WRITE_BYTE(TE_FIREFIELD);
+		WRITE_COORD(position[0]);
+		WRITE_COORD(position[1]);
+		WRITE_COORD(position[2]);
+		WRITE_SHORT(params[arg_radius]);
+		WRITE_SHORT(params[arg_sprite]);
+		WRITE_BYTE(params[arg_count]);
+		WRITE_BYTE(params[arg_flags]);
+		WRITE_BYTE(params[arg_duration]);
+	MESSAGE_END();
+
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL te_attach_model_to_player(AMX *amx, cell *params)
+{
+	enum args { arg_numargs, arg_player, arg_model, arg_offset, arg_life, arg_receiver, arg_reliable };
+
+	int index = params[arg_receiver];
+
+	if(!check_msg_receiver(amx, index))
+		return 0;
+
+	MESSAGE_BEGIN(get_msg_destination(index, params[arg_reliable] != 0), SVC_TEMPENTITY, NULL, index ? TypeConversion.id_to_edict(index) : NULL);
+		WRITE_BYTE(TE_PLAYERATTACHMENT);
+		WRITE_BYTE(params[arg_player]);
+		WRITE_COORD(params[arg_offset]);
+		WRITE_SHORT(params[arg_model]);
+		WRITE_SHORT(params[arg_life]);
+	MESSAGE_END();
+
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL te_remove_all_player_attachments(AMX *amx, cell *params)
+{
+	enum args { arg_numargs, arg_player, arg_receiver, arg_reliable };
+
+	int index = params[arg_receiver];
+
+	if(!check_msg_receiver(amx, index))
+		return 0;
+
+	MESSAGE_BEGIN(get_msg_destination(index, params[arg_reliable] != 0), SVC_TEMPENTITY, NULL, index ? TypeConversion.id_to_edict(index) : NULL);
+		WRITE_BYTE(TE_KILLPLAYERATTACHMENTS);
+		WRITE_BYTE(params[arg_player]);
+	MESSAGE_END();
+
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL te_create_multi_gunshot(AMX *amx, cell *params)
+{
+	enum args { arg_numargs, arg_position, arg_direction, arg_decal, arg_count, arg_noise_x, arg_noise_y, arg_receiver, arg_reliable };
+
+	int index = params[arg_receiver];
+
+	if(!check_msg_receiver(amx, index))
+		return 0;
+
+	cell *position = get_amxaddr(amx, params[arg_position]);
+	cell *direction = get_amxaddr(amx, params[arg_direction]);
+
+	MESSAGE_BEGIN(get_msg_destination(index, params[arg_reliable] != 0), SVC_TEMPENTITY, NULL, index ? TypeConversion.id_to_edict(index) : NULL);
+		WRITE_BYTE(TE_MULTIGUNSHOT);
+		WRITE_COORD(position[0]);
+		WRITE_COORD(position[1]);
+		WRITE_COORD(position[2]);
+		WRITE_COORD(direction[0]);
+		WRITE_COORD(direction[1]);
+		WRITE_COORD(direction[2]);
+		WRITE_COORD(params[arg_noise_x]);
+		WRITE_COORD(params[arg_noise_y]);
+		WRITE_BYTE(params[arg_count]);
+		WRITE_BYTE(params[arg_decal]);
+	MESSAGE_END();
+
+	return 1;
+}
+
+static cell AMX_NATIVE_CALL te_create_user_tracer(AMX *amx, cell *params)
+{
+	enum args { arg_numargs, arg_position, arg_velocity, arg_life, arg_color, arg_length, arg_receiver, arg_reliable };
+
+	int index = params[arg_receiver];
+
+	if(!check_msg_receiver(amx, index))
+		return 0;
+
+	// Values outside the range 0-12 will throw an error
```

Comment:
So... why it isn't thrown here?

Reply:
@WPMGPRoSToTeMa I don't think it has to. Since all other natives accept values from 0 to 255, this one being the only one limited from 0 to 12 will probably get forgotten by users, so better use `clamp` to limit it automatically.

Reply:
@OciXCrom there is another way - we can allow to put RGB and then choose the nearest color to it.

Reply:
@WPMGPRoSToTeMa That would be cool if it would work correctly. Can you show an example?

Reply:
@OciXCrom I think it is better to just add a tagged enum for `color` param, it is more explicit and safe.

Reply:
@WPMGPRoSToTeMa Good idea, I added them. However I don't see any visible difference between 6, 8, 9, 10 and 11, so I named all of them `TracerColor_Pink%num%`.

Reply:
@OciXCrom here the list of colors:
```
0 - FFFFFF
1 - FF0000
2 - 00FF00
3 - 0000FF
4 - 000000
5 - FFA711
6 - FF825A
7 - 373C90
8 - FF825A
9 - FF8C5A
10 - C8825A
11 - FF7846
12 - 000000
```
You can convert them to color names using [this](https://www.color-blindness.com/color-name-hue/) (for example).

Reply:
@WPMGPRoSToTeMa Are you sure this list is correct? The 12th color definitely isn't #000000.

Reply:
@OciXCrom answered here https://github.com/alliedmodders/amxmodx/pull/523#discussion_r214147046.